### ISSUE TITLE
rewriter tests fail/crash on ppc64

### ISSUE THREAD

**Issue by wrwilliams** [dyninst/dyninst#34]
Something is systemically wrong with rewriting on this platform.

**Comment by wrwilliams** [comment]
The tramp guard locking functions we're executing are junk; the ones I get from disassembling the RTlib look good.

**Comment by jdetter** [comment]
@wrwilliams was this fixed for ppc64? I think we fixed a similar issue on x86_64.

**Comment by wrwilliams** [comment]
I haven't checked recently, but that's on today's list.

**Comment by wrwilliams** [comment]
There's TOC glue code that doesn't line up with what we expect for either PPC ABI (v2 we'd expect that the local and global entries are contiguous and they're not, v1 we'd expect that there's not two separate entry points and there are). I think this is a won't fix until we work out the full implications of PPC ABI v2; removing from milestone but leaving open.

**Comment by mcfadden8** [comment]
Hi @wrwilliams, 

for ABIv2, what is an example of when local and global entries are not contiguous?

**Comment by wrwilliams** [comment]
...I'm not 100% positive what I'm seeing, but here are the relevant snippets from our runtime library, which exports DYNINST_unlock_tramp_guard:

```0000000000003608 <00000010.plt_call.DYNINST_unlock_tramp_guard+0>:
    3608:       f8 41 00 28     std     r2,40(r1)
    360c:       e9 62 88 90     ld      r11,-30576(r2)
    3610:       7d 69 03 a6     mtctr   r11
    3614:       e8 42 88 98     ld      r2,-30568(r2)
    3618:       28 22 00 00     cmpldi  r2,0
    361c:       4c e2 04 20     .long 0x4ce20420
    3620:       48 00 35 80     b       6ba0 <DYNINST_unlock_tramp_guard@plt>```

which is obviously a thing doing TOC setup of some sort, and:

```0000000000003b70 <.DYNINST_unlock_tramp_guard>:
    3b70:       7c 08 02 a6     mflr    r0
    3b74:       f8 01 00 10     std     r0,16(r1)
    3b78:       f8 21 ff 91     stdu    r1,-112(r1)
    3b7c:       38 62 81 58     addi    r3,r2,-32424
    3b80:       4b ff fb bd     bl      373c <00000010.plt_call.__tls_get_addr@@GLIBC_2.3+0>
    3b84:       e8 41 00 28     ld      r2,40(r1)
    3b88:       38 21 00 70     addi    r1,r1,112
    3b8c:       38 00 00 01     li      r0,1
    3b90:       90 03 00 00     stw     r0,0(r3)
    3b94:       e8 01 00 10     ld      r0,16(r1)
    3b98:       7c 08 03 a6     mtlr    r0
    3b9c:       4e 80 00 20     blr```

which is the actual function.

It looks like we're generating what would be a proper intermodule call for ABIv1 in our instrumentation, and we should be sending it to the actual function, which *should* be okay.

```101001ec:   3d 80 00 0c     lis     r12,12
    101001f0:   61 8c d4 30     ori     r12,r12,54320
    101001f4:   7d 8c 12 14     add     r12,r12,r2
    101001f8:   e8 4c 00 08     ld      r2,8(r12)
    101001fc:   e9 8c 00 00     ld      r12,0(r12)
    10100200:   7d 88 03 a6     mtlr    r12
    10100204:   e9 81 00 18     ld      r12,24(r1)
    10100208:   4e 80 00 21     blrl
    1010020c:   e8 41 00 28     ld      r2,40(r1)
    10100210:   60 6c 00 00     ori     r12,r3,0
    10100214:   e8 41 00 c0     ld      r2,192(r1)
    10100218:   2c 0c 00 00     cmpwi   r12,0
    1010021c:   41 82 00 4c     beq-    10100268 <_ZL12eventCounter+0xd3180>
    10100220:   4b f0 77 35     bl      10007954 <targ10006174+0x17e0>```

(the symbols are messed up but I don't think that's actually the issue)

When we hit the call at the end of the above, gdb says we're in DYNINST_unlock_tramp_guard, but that the disassembly is complete garbage (and doesn't match the snippet above from disassembling the runtime).

Thoughts?

**Comment by cuviper** [comment]
Have you tried this again since the static TLS #99 went in?  That should get rid of the `__tls_get_addr` call at least, but I have no idea if that's any part of your problem.

**Comment by wrwilliams** [comment]
Testing now, but an update for @mcfadden8 and @mplegendre:

The runtime library snippets above are generated by /usr/bin/cc on vulcan, which claims to be gcc 4.4.7 when run with -v:

```
vulcanlac4{bill}37: /usr/bin/cc -v
Using built-in specs.
Target: ppc64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-languages=c,c++,objc,obj-c++,java,fortran,ada --enable-java-awt=gtk --disable-dssi --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-1.5.0.0/jre --enable-libgcj-multifile --enable-java-maintainer-mode --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libjava-multilib --with-ppl --with-cloog --enable-secureplt --with-long-double-128 --with-cpu-32=power4 --with-tune-32=power6 --with-cpu-64=power4 --with-tune-64=power6 --build=ppc64-redhat-linux
Thread model: posix
gcc version 4.4.7 20120313 (Red Hat 4.4.7-11) (GCC) 
```

**Comment by wrwilliams** [comment]
No change in behavior with #99, no regression in behavior with #109--which makes sense, as we're not even getting to execute DYNINST_unlock_tramp_guard before we crash. The actual call to unlock is what's busted.

**Comment by mcfadden8** [comment]
@wrwilliams, which code fragment above seems to be ABIv2?

**Comment by wrwilliams** [comment]
The first two match up by semantics, but not by name or layout, to the global/local entry point behavior you've described:

Global entry point, in libdyninstAPI_RT.so:
```
0000000000003608 <00000010.plt_call.DYNINST_unlock_tramp_guard+0>:
3608: f8 41 00 28 std r2,40(r1)
360c: e9 62 88 90 ld r11,-30576(r2)
3610: 7d 69 03 a6 mtctr r11
3614: e8 42 88 98 ld r2,-30568(r2)
3618: 28 22 00 00 cmpldi r2,0
361c: 4c e2 04 20 .long 0x4ce20420
3620: 48 00 35 80 b 6ba0 DYNINST_unlock_tramp_guard@plt
```
Local entry point, also in libdyninstAPI_RT.so, but not contiguous:
```
0000000000003b70 <.DYNINST_unlock_tramp_guard>:
3b70: 7c 08 02 a6 mflr r0
3b74: f8 01 00 10 std r0,16(r1)
3b78: f8 21 ff 91 stdu r1,-112(r1)
3b7c: 38 62 81 58 addi r3,r2,-32424
3b80: 4b ff fb bd bl 373c <00000010.plt_call.__tls_get_addr@@GLIBC_2.3+0>
3b84: e8 41 00 28 ld r2,40(r1)
3b88: 38 21 00 70 addi r1,r1,112
3b8c: 38 00 00 01 li r0,1
3b90: 90 03 00 00 stw r0,0(r3)
3b94: e8 01 00 10 ld r0,16(r1)
3b98: 7c 08 03 a6 mtlr r0
3b9c: 4e 80 00 20 blr
```

Of course I'm also not sure how there's any PLT nonsense that should be associated with this at all, but note that the intermodule call out of the snippet at 3b70 is calling something with name mangling like that at 3608, and is (assuming I'm reading the assembly correctly) putting a TOC-derived value in r3 but not touching the TOC register itself.

Am I misremembering where and how the v1 ABI did its TOC munging? I really thought that was 100% caller-side...

**Comment by mcfadden8** [comment]
What I had previously described for global/local entry point behavior was for ABIv2.

This all seems to be ABIv1. I don't know too much about ABIv1, but if I understand correctly what I am reading in the specification, there are three potential points of entry:

1. A linkage function that is set up by the link editor (which I think is shown first)
2. A function descriptor (which I do not think is shown but is referenced (no "."))
3. The function entry which I think is the third fragment

**Comment by mcfadden8** [comment]
Hi @wrwilliams, 

In the generated code:
``` assembly
  1 101001ec: 3d 80 00 0c lis r12,12
  2 101001f0: 61 8c d4 30 ori r12,r12,54320 
  3 101001f4: 7d 8c 12 14 add r12,r12,r2
  4 101001f8: e8 4c 00 08 ld r2,8(r12)
  5 101001fc: e9 8c 00 00 ld r12,0(r12)
  6 10100200: 7d 88 03 a6 mtlr r12
  7 10100204: e9 81 00 18 ld r12,24(r1)
  8 10100208: 4e 80 00 21 blrl
  9 1010020c: e8 41 00 28 ld r2,40(r1)
 10 10100210: 60 6c 00 00 ori r12,r3,0
 11 10100214: e8 41 00 c0 ld r2,192(r1)
 12 10100218: 2c 0c 00 00 cmpwi r12,0
 13 1010021c: 41 82 00 4c beq- 10100268 <_ZL12eventCounter+0xd3180>
 14 10100220: 4b f0 77 35 bl 10007954
```
Is the TOC what you expected it to be at the end of this code? It appears that line 9 is correctly restoring the TOC after the call, but I am not sure what line 11 doing?

**Comment by wrwilliams** [comment]
Fixed by #156 per @cuviper
